File: Front-End/meu-fontEnd/src/services/gastosService.ts
import { financeApi } from "./api"
import type { criarGastos, Gastos } from "../types/Gastos"

// LISTAR GASTOS
export const listarGastos = async (): Promise<Gastos[]> => {
  const { data } = await financeApi.get("/gastos")
  return data
}

// CRIAR GASTO
export const novoGasto = async (dados: criarGastos): Promise<Gastos> => {
  const payload = {
    descricao: dados.descricao,
    valor: Number(dados.valor),
    data: dados.data,
    categoria: dados.categoria,
  }

  const { data } = await financeApi.post("/gastos", payload)
  return data
}

// DELETAR GASTO
export const deletarGasto = async (id: number): Promise<void> => {
  await financeApi.delete(`/gastos/${id}`)
}

// ATUALIZAR GASTO
export const atualizarGasto = async (
  id: number,
  dados: criarGastos
): Promise<Gastos> => {
  const payload = {
    descricao: dados.descricao,
    valor: Number(dados.valor),
    data: dados.data,
    categoria: dados.categoria,
  }

  const { data } = await financeApi.put(`/gastos/${id}`, payload)
  return data
}

File: privado/src/main/java/financeiro/privado/service/GastosService.java
package financeiro.privado.service;

import financeiro.privado.dto.GastosRequest;
import financeiro.privado.model.Gastos;
import financeiro.privado.repository.GastosRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;

import java.util.List;
import java.util.stream.Collectors;

@Service
@RequiredArgsConstructor
public class GastosService {

    private final GastosRepository gastosRepository;

    public Gastos novoGastos(GastosRequest gastosRequest, Long usuarioId) {
        Gastos novoGasto = Gastos.builder()
                .valor(gastosRequest.getValor())
                .descricao(gastosRequest.getDescricao())
                .data(gastosRequest.getData())
                .categoria(gastosRequest.getCategoria())
                .usuarioId(usuarioId)
                .build();

        Gastos gastoSalvo = gastosRepository.save(novoGasto);

        return transformarParaResponse(gastoSalvo);
    }

    private Gastos transformarParaResponse(Gastos g) {
        return Gastos.builder()
                .id(g.getId())
                .data(g.getData())
                .valor(g.getValor())
                .descricao(g.getDescricao())
                .categoria(g.getCategoria())
                .usuarioId(g.getUsuarioId())
                .build();
    }

    public List<Gastos> listarGastos(Long usuarioId) {
        List<Gastos> gastos = gastosRepository.findByUsuarioId(usuarioId);

        return gastos.stream()
                .map(this::transformarParaResponse)
                .collect(Collectors.toList());
    }

    public void deletarGastos(Long id, Long usuarioId) {
        Gastos gasto = gastosRepository.findByIdAndUsuarioId(id, usuarioId)
                .orElseThrow(() -> new RuntimeException("TRANSAÇÃO NÃO ENCONTRADA COM O ID: " + id));

        gastosRepository.delete(gasto);
    }

    public Gastos atualizarGasto(Long id, GastosRequest gastosRequest, Long usuarioId) {
        Gastos gastosExistente = gastosRepository.findByIdAndUsuarioId(id, usuarioId)
                .orElseThrow(() -> new RuntimeException("TRANSAÇÃO NÃO ENCONTRADA"));

        gastosExistente.setDescricao(gastosRequest.getDescricao());
        gastosExistente.setValor(gastosRequest.getValor());
        gastosExistente.setCategoria(gastosRequest.getCategoria());
        gastosExistente.setData(gastosRequest.getData());

        Gastos gastosAtualizado = gastosRepository.save(gastosExistente);

        return transformarParaResponse(gastosAtualizado);
    }
}